plugins/catalog-backend-module-bitbucket-server/src/providers/BitbucketServerEntityProvider.ts

```diff
@@ -218,10 +261,251 @@ export class BitbucketServerEntityProvider implements EntityProvider {
+    if (!this.canHandleEvents()) {
+      this.logger.error(
+        'Bitbucket Server catalog entity provider is not set up to handle events. Missing authService or catalogApi.',
+      );
+      return;
+    }
+
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const repoSlug = event.repository.slug;
+    const catalogRepoUrl: string = `https://${this.config.host}/projects/${event.repository.project.key}/repos/${repoSlug}/browse${this.config.catalogPath}`;
+    this.logger.info(`handle repo:push event for ${catalogRepoUrl}`);
+    const targets = await this.getLocationEntity(event);
+    if (targets.length === 0) {
+      this.logger.error('Failed to create location entity.');
+      return;
+    }
+    const { token } = await this.auth!.getPluginRequestToken({
+      onBehalfOf: await this.auth!.getOwnServiceCredentials(),
+      targetPluginId: 'catalog', // e.g. 'catalog'
+    });
+    const existing = await this.findExistingLocations(catalogRepoUrl, token);
+    const stillExisting: LocationEntity[] = [];
+    const removed: DeferredEntity[] = [];
+    existing.forEach(item => {
+      if (
+        targets.find(
+          value =>
+            value.metadata.annotations![this.targetAnnotation] ===
+            item.spec.target,
+        )
+      ) {
+        stillExisting.push(item);
+      } else {
+        removed.push({
+          locationKey: this.getProviderName(),
+          entity: item,
+        });
+      }
+    });
+
+    const added = await this.getAddedEntities(targets, existing);
+
+    if (
+      stillExisting.length > 0 &&
+      stillExisting[0].metadata.annotations![this.defaultBranchAnnotation] !==
+        undefined &&
+      !this.isDefaultBranchPush(
+        stillExisting[0].metadata.annotations![this.defaultBranchAnnotation],
+        event,
+      )
+    ) {
+      return;
+    } else if (
+      added.length > 0 &&
+      added[0].entity.metadata.annotations![this.defaultBranchAnnotation] !==
+        undefined &&
+      !this.isDefaultBranchPush(
+        added[0].entity.metadata.annotations![this.defaultBranchAnnotation],
+        event,
+      )
+    ) {
+      return;
+    } else if (
+      removed.length > 0 &&
+      removed[0].entity.metadata.annotations![this.defaultBranchAnnotation] !==
+        undefined &&
+      !this.isDefaultBranchPush(
+        removed[0].entity.metadata.annotations![this.defaultBranchAnnotation],
+        event,
+      )
+    ) {
+      return;
+    }
+
+    const promises: Promise<void>[] = [
+      this.connection.refresh({
+        keys: stillExisting.map(entity => `url:${entity.spec.target}`),
+      }),
+    ];
+
+    if (added.length > 0 || removed.length > 0) {
+      promises.push(
+        this.connection.applyMutation({
+          type: 'delta',
+          added: added,
+          removed: removed,
+        }),
+      );
+    }
+
+    await Promise.all(promises);
+
+    return;
+  }
+
+  private async getAddedEntities(
```

Comment:
Hm is it useful to note that `added` here actually means `upsert`? It's fine to send "added" delta events with things that actually existed beforehand and which does NOT have changes. Even if it's wasteful. It's much better to do that then to call out to the catalog to try to make a diff - because the catalog already performs that diff internally anyway and throws away no-op things very efficiently.

But maybe you need this anyway to be able to detect deletions. Then all good